wrappers/icons.go

```diff
@@ -0,0 +1,125 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package wrappers
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+)
+
+func findIconFiles(s *snap.Info, rootDir string) (icons []string, err error) {
+	if !osutil.IsDirectory(rootDir) {
+		return nil, nil
+	}
+	iconGlob := fmt.Sprintf("snap.%s.*", s.SnapName())
+	forbiddenDirGlob := "snap.*"
+	err = filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		rel, err := filepath.Rel(rootDir, path)
+		if err != nil {
+			return err
+		}
+		base := filepath.Base(path)
+		if info.IsDir() {
+			// Ignore directories that could match an icon glob
+			if ok, err := filepath.Match(forbiddenDirGlob, base); ok || err != nil {
+				return filepath.SkipDir
+			}
+		} else {
+			if ok, err := filepath.Match(iconGlob, base); err != nil {
+				return err
+			} else if ok {
+				ext := filepath.Ext(base)
+				if ext == ".png" || ext == ".svg" {
+					icons = append(icons, rel)
+				}
+			}
+		}
+		return nil
+	})
+	return icons, err
+}
+
+func deriveIconContent(s *snap.Info, rootDir string, icons []string) (content map[string]map[string]*osutil.FileState, err error) {
+	content = make(map[string]map[string]*osutil.FileState)
+	snapPrefix := fmt.Sprintf("snap.%s.", s.SnapName())
+	instancePrefix := fmt.Sprintf("snap.%s.", s.InstanceName())
+
+	for _, iconFile := range icons {
+		dir := filepath.Dir(iconFile)
+		base := filepath.Base(iconFile)
+		dirContent := content[dir]
+		if dirContent == nil {
+			dirContent = make(map[string]*osutil.FileState)
+			content[dir] = dirContent
+		}
+		data, err := ioutil.ReadFile(filepath.Join(rootDir, iconFile))
```

Comment:
should we check at least that icons haven't an unreasonable size?

Reply:
Probably a good idea.  I wonder what we should consider unreasonable?

Looking at the icons in `/usr/share/icons` on my system, the largest ones seem to be Xcursor format files at about 4MB, which seem to be animated mouse cursors.  For actual application icons, Handbrake seems to be an outlier with 3.5MB SVG icon (which seems to contain a large embedded base64 encoded PNG file).

Ignoring those outliers, there are a number of PNG and SVG icons between 100KB-140KB, so "reasonable" needs to include those sizes.

Reply:
If we put a limit (8MB ?)  it probably needs to be checked already in snap/validate.go so that is not a surprise only at installation.

Another option is to have Source path field in FileState, if it's set and the source is over some threshold we do a buffered file comparison and copy instead of the whole to memory approach.

Both these seems more follow up material than something to do here though.

Reply:
As a separate idea, we can extend `EnsureDirState` to support streaming. Instead of handing out bytes we could hand out bytes or file references that it can then efficiently use to avoid holding the entire file in memory.

Reply:
> Yeah, I'd like to second this request. I think it can merge but with my virtual-seciruty-hat on I'd say this is a way to DOS snapd. Just put up a large enough .png file full of noise and snapd goes down.
> 
> CC @jdstrand for possible validation angle at the store.

I'm working on this in the review-tools now. Please don't block this PR on that.